src/huggingface_hub/file_download.py

```diff
@@ -366,42 +366,87 @@ def _raise_if_offline_mode_is_enabled(msg: Optional[str] = None):
         )
 
 
-def _request_with_retry(
+def _request_wrapper(
     method: str,
     url: str,
     *,
     max_retries: int = 0,
     base_wait_time: float = 0.5,
     max_wait_time: float = 2,
     timeout: float = 10.0,
+    force_internal_redirects: bool = False,
     **params,
 ) -> requests.Response:
-    """Wrapper around requests to retry in case it fails with a `ConnectTimeout`, with
-    exponential backoff.
-
-        Note that if the environment variable HF_HUB_OFFLINE is set to 1, then a
-        `OfflineModeIsEnabled` error is raised.
-
-        Args:
-            method (`str`):
-                HTTP method, such as 'GET' or 'HEAD'.
-            url (`str`):
-                The URL of the resource to fetch.
-            max_retries (`int`, *optional*, defaults to `0`):
-                Maximum number of retries, defaults to 0 (no retries).
-            base_wait_time (`float`, *optional*, defaults to `0.5`):
-                Duration (in seconds) to wait before retrying the first time.
-                Wait time between retries then grows exponentially, capped by
-                `max_wait_time`.
-            max_wait_time (`float`, *optional*, defaults to `2`):
-                Maximum amount of time between two retries, in seconds.
-            timeout (`float`, *optional*, defaults to `10`):
-                How many seconds to wait for the server to send data before
-                giving up which is passed to `requests.request`.
-            **params (`dict`, *optional*):
-                Params to pass to `requests.request`.
+    """Wrapper around requests methods to add several features.
+
+    What it does:
+    1. Ensure offline mode is disabled (env variable `HF_HUB_OFFLINE` not set to 1).
+       If enabled, a `OfflineModeIsEnabled` exception is raised.
+    2. Handle internal redirection (if `force_internal_redirects=True`).
+    3. Retry in case request fails with a `ConnectTimeout`, with exponential backoff.
+
+    Args:
+        method (`str`):
+            HTTP method, such as 'GET' or 'HEAD'.
+        url (`str`):
+            The URL of the resource to fetch.
+        max_retries (`int`, *optional*, defaults to `0`):
+            Maximum number of retries, defaults to 0 (no retries).
+        base_wait_time (`float`, *optional*, defaults to `0.5`):
+            Duration (in seconds) to wait before retrying the first time.
+            Wait time between retries then grows exponentially, capped by
+            `max_wait_time`.
+        max_wait_time (`float`, *optional*, defaults to `2`):
+            Maximum amount of time between two retries, in seconds.
+        timeout (`float`, *optional*, defaults to `10`):
+            How many seconds to wait for the server to send data before
+            giving up which is passed to `requests.request`.
+        force_internal_redirects (`bool`, *optional*, defaults to `False`)
```

Comment:
Maybe `follow_relative_redirects` or `follow_same_site_redirects` would be clearer

Reply:
Also note that `requests` follows redirects by default, except for HEAD requests (see their docs: https://requests.readthedocs.io/en/latest/user/quickstart/#redirection-and-history)

Reply:
> Also note that requests follows redirects by default, except for HEAD requests (see their docs: https://requests.readthedocs.io/en/latest/user/quickstart/#redirection-and-history)

Yep exactly + we specifically set `allow_redirects` to False in `hf_hub_download` to be sure we don't follow redirection to the CDN for lfs files. So here we want to block redirections except relative ones.